File: frontend/src/types/contest.ts
/**
 * 比赛与挑战题目相关类型定义。
 * 这里统一承载解题赛、对抗赛、观战和回放相关的领域契约。
 */

import type { BattleOrchestration, ChallengeOrchestration } from './contestOrchestration'

export type ContestType = 'jeopardy' | 'agent_battle'
export type ContestLevel = 'practice' | 'school' | 'cross_school' | 'platform'
export type ContestStatus = 'draft' | 'published' | 'ongoing' | 'ended'

export type ChallengeDifficulty = 1 | 2 | 3 | 4 | 5

export type ChallengeCategory =
  | 'contract_vuln'
  | 'defi'
  | 'consensus'
  | 'crypto'
  | 'cross_chain'
  | 'nft'
  | 'reverse'
  | 'key_management'
  | 'misc'

export type ChallengeRuntimeProfile =
  | 'static'
  | 'single_chain_instance'
  | 'fork_replay'
  | 'multi_service_lab'

export type FlagType = 'static' | 'dynamic'
export type ChallengeSourceType = 'preset' | 'auto_converted' | 'user_created'
export type ChallengeStatus = 'draft' | 'active' | 'archived'

export interface Contest {
  id: number
  school_id?: number
  creator_id: number
  creator_name?: string
  title: string
  name?: string
  description?: string
  type: ContestType
  level?: ContestLevel
  cover?: string
  rules?: string
  start_time: string
  end_time: string
  registration_start?: string
  registration_end?: string
  dynamic_score?: boolean
  first_blood_bonus?: number
  max_participants?: number
  team_min_size?: number
  team_max_size?: number
  battle_orchestration?: BattleOrchestration
  status: ContestStatus
  is_public?: boolean
  created_at: string
  participant_count?: number
  challenge_count?: number
  is_registered?: boolean
}

export interface Challenge {
  id: number
  creator_id: number
  creator_name?: string
  school_id?: number
  title: string
  description: string
  category: ChallengeCategory
  runtime_profile: ChallengeRuntimeProfile
  difficulty: ChallengeDifficulty
  base_points: number
  points?: number
  min_points?: number
  decay_factor?: number
  flag_type: FlagType
  flag_template?: string
  contract_code?: string
  setup_code?: string
  deploy_script?: string
  check_script?: string
  validation_config?: Record<string, unknown>
  challenge_orchestration: ChallengeOrchestration
  hints?: ChallengeHint[]
  attachments?: string[]
  tags?: string[]
  source_type?: ChallengeSourceType
  source_ref?: string
  status: ChallengeStatus
  is_public: boolean
  solve_count?: number
  attempt_count?: number
  created_at: string
  is_solved?: boolean
  awarded_points?: number
  first_blood?: string
  first_blood_time?: string
}

export interface ChallengeHint {
  content: string
  cost: number
}

export interface CreateChallengeRequest {
  title: string
  description: string
  category: ChallengeCategory
  runtime_profile: ChallengeRuntimeProfile
  difficulty: number
  base_points: number
  min_points?: number
  decay_factor?: number
  flag_type: FlagType
  flag_template?: string
  contract_code?: string
  setup_code?: string
  deploy_script?: string
  check_script?: string
  challenge_orchestration: ChallengeOrchestration
  hints?: ChallengeHint[]
  attachments?: string[]
  tags?: string[]
  is_public: boolean
}

export interface UpdateChallengeRequest extends Partial<CreateChallengeRequest> {
  status?: ChallengeStatus
}

export interface ContestChallenge {
  id: number
  contest_id: number
  challenge_id: number
  points: number
  current_points: number
  sort_order: number
  is_visible: boolean
  challenge: Challenge
  is_solved?: boolean
  solve_count?: number
}

export interface Team {
  id: number
  contest_id: number
  name: string
  captain_id: number
  leader_name?: string
  avatar?: string
  description?: string
  status: 'forming' | 'ready' | 'competing'
  invite_code?: string
  member_count?: number
  members?: TeamMember[]
  created_at: string
}

export interface TeamMember {
  id: number
  team_id: number
  user_id: number
  display_name?: string
  real_name?: string
  phone?: string
  student_no?: string
  avatar?: string
  role: 'captain' | 'member'
  is_captain?: boolean
  joined_at: string
}

export interface ContestScore {
  rank: number
  user_id?: number
  display_name?: string
  team_id?: number
  team_name?: string
  total_score: number
  solve_count: number
  last_solve_at?: string
  first_blood_count?: number
}

export interface Scoreboard {
  list: ContestScore[]
  my_rank?: number
  my_score?: number
}

export interface AgentBattleScore {
  rank: number
  team_id: number
  team_name: string
  score: number
  token_balance: string
  success_count: number
  fail_count: number
  resource_held: number
}

export interface ChallengeEnv {
  id: number
  env_id: string
  contest_id: number
  challenge_id: number
  status: string
  access_url?: string
  tools?: Array<{
    key: string
    label: string
    kind?: string
    target?: string
    port: number
    route: string
    instance_route?: string
  }>
  service_entries?: Array<{
    key: string
    label: string
    description?: string
    purpose?: string
    access_url?: string
    protocol?: string
    port?: number
    expose_as?: string
  }>
  started_at?: string
  expires_at?: string
  error_message?: string
  remaining: number
}

export interface CategoryGroup {
  category: string
  label: string
  challenges: Challenge[]
}

export interface TeamContractInfo {
  id: number
  contract_address: string
  status: string
  version: number
  deployed_at?: string
}

export type BattleRoundPhase =
  | 'pending'
  | 'upgrade_window'
  | 'locked'
  | 'executing'
  | 'settling'
  | 'finished'

export interface CurrentRoundInfo {
  id: number
  round_number: number
  status: string
  phase?: BattleRoundPhase | string
  start_time?: string
  end_time?: string
  upgrade_window_end?: string
}

export interface FlagSubmitResult {
  correct: boolean
  points: number
  message: string
}

export interface AgentBattleRound {
  round_number: number
  start_time: string
  end_time: string
  upgrade_window_end?: string
  status: string
  phase?: BattleRoundPhase | string
}

export interface AgentBattleTeamStatus {
  team_id: number
  team_name: string
  contract_address?: string
  is_alive: boolean
  resource_held: number
  total_score: number
  score_change?: string
}

export interface AgentBattleEvent {
  block: number
  round_number?: number
  time?: string
  event_type: string
  actor_team?: string
  target_team?: string
  action_result?: string
  score_delta?: number
  resource_delta?: number
  points?: number
  description?: string
}

export type BattleStatus = 'waiting' | 'running' | 'paused' | 'completed'

export interface AgentBattleStatus {
  status?: BattleStatus
  current_block?: number
  current_round?: number
  total_rounds?: number
  round_phase?: BattleRoundPhase | string
  my_rank?: number
  my_score?: number
  teams?: AgentBattleTeamStatus[]
  recent_events?: AgentBattleEvent[]
  agent_status?: {
    version: string
    uploaded_at: string
    is_valid: boolean
  }
}

export interface SpectateData {
  contest_name: string
  current_round: number
  current_block?: number
  round_status: string
  round_phase?: BattleRoundPhase | string
  round_end_time?: string
  teams?: AgentBattleTeamStatus[]
  recent_events?: AgentBattleEvent[]
  spectator_count?: number
  available_battles?: Array<{ id: string; name: string }>
  current_battle?: {
    round: number
    red_team: string
    blue_team: string
    current_turn: number
  }
  visualization_data?: Record<string, unknown>
}

export const ContestTypeMap: Record<ContestType, string> = {
  jeopardy: '解题赛',
  agent_battle: '智能体博弈战',
}

export const ContestLevelMap: Record<ContestLevel, string> = {
  practice: '练习赛',
  school: '校内赛',
  cross_school: '跨校赛',
  platform: '平台赛',
}

export const ContestStatusMap: Record<ContestStatus, { text: string; color: string }> = {
  draft: { text: '草稿', color: 'default' },
  published: { text: '已发布', color: 'processing' },
  ongoing: { text: '进行中', color: 'success' },
  ended: { text: '已结束', color: 'default' },
}

export const ChallengeRuntimeProfileMap: Record<ChallengeRuntimeProfile, string> = {
  static: '静态题',
  single_chain_instance: '链上实例题',
  fork_replay: 'Fork 复现题',
  multi_service_lab: '多服务拓扑题',
}

export const DifficultyMap: Record<string | number, { text: string; color: string; stars: number }> = {
  1: { text: '入门', color: 'blue', stars: 1 },
  2: { text: '简单', color: 'green', stars: 2 },
  3: { text: '中等', color: 'orange', stars: 3 },
  4: { text: '困难', color: 'red', stars: 4 },
  5: { text: '专家', color: 'purple', stars: 5 },
  easy: { text: '简单', color: 'green', stars: 1 },
  medium: { text: '中等', color: 'orange', stars: 2 },
  hard: { text: '困难', color: 'red', stars: 3 },
  expert: { text: '专家', color: 'purple', stars: 4 },
}

export const CategoryMap: Record<string, string> = {
  contract_vuln: '合约漏洞',
  defi: 'DeFi',
  consensus: '共识机制',
  crypto: '密码学',
  cross_chain: '跨链安全',
  nft: 'NFT/代币安全',
  reverse: '逆向分析',
  key_management: '密钥安全',
  misc: '其他',
}

export const BattleStatusMap: Record<BattleStatus, { text: string; color: string }> = {
  waiting: { text: '等待中', color: 'default' },
  running: { text: '进行中', color: 'success' },
  paused: { text: '已暂停', color: 'warning' },
  completed: { text: '已完成', color: 'blue' },
}

export const BattleRoundPhaseMap: Record<BattleRoundPhase, { text: string; color: string }> = {
  pending: { text: '待开始', color: 'default' },
  upgrade_window: { text: '升级窗口', color: 'processing' },
  locked: { text: '锁定阶段', color: 'warning' },
  executing: { text: '执行阶段', color: 'success' },
  settling: { text: '结算阶段', color: 'purple' },
  finished: { text: '已结束', color: 'default' },
}

export interface CreateContestRequest {
  title: string
  description?: string
  type: ContestType
  level?: string
  cover?: string
  rules?: string
  start_time: string
  end_time: string
  registration_start?: string
  registration_end?: string
  dynamic_score?: boolean
  first_blood_bonus?: number
  max_participants?: number
  team_min_size?: number
  team_max_size?: number
  is_public?: boolean
  battle_orchestration?: BattleOrchestration
}

export interface ContestRecord {
  contest_id: number
  contest_name: string
  contest_type: string
  team_name?: string
  rank?: number
  total_score: number
  status: ContestStatus
}

File: backend/internal/dto/request/contest.go
package request

import "github.com/chainspace/backend/internal/model"

type ChallengeHintPayload struct {
	Content string `json:"content" binding:"required"`
	Cost    int    `json:"cost" binding:"min=0"`
}

// CreateContestRequest 创建竞赛请求
type CreateContestRequest struct {
	Level               string                    `json:"level" binding:"omitempty,oneof=practice school cross_school platform"`
	Title               string                    `json:"title" binding:"required,max=200"`
	Description         string                    `json:"description" binding:"omitempty"`
	Type                string                    `json:"type" binding:"required,oneof=jeopardy agent_battle"`
	Cover               string                    `json:"cover" binding:"omitempty,max=500"`
	Rules               string                    `json:"rules" binding:"omitempty"`
	StartTime           string                    `json:"start_time" binding:"required"`
	EndTime             string                    `json:"end_time" binding:"required"`
	RegistrationStart   string                    `json:"registration_start" binding:"omitempty"`
	RegistrationEnd     string                    `json:"registration_end" binding:"omitempty"`
	IsPublic            bool                      `json:"is_public"`
	MaxParticipants     int                       `json:"max_participants" binding:"omitempty,min=0"`
	TeamMinSize         int                       `json:"team_min_size" binding:"omitempty,min=1"`
	TeamMaxSize         int                       `json:"team_max_size" binding:"omitempty,min=1"`
	DynamicScore        bool                      `json:"dynamic_score"`
	FirstBloodBonus     int                       `json:"first_blood_bonus" binding:"omitempty,min=0"`
	BattleOrchestration model.BattleOrchestration `json:"battle_orchestration"`
}

// UpdateContestRequest 更新竞赛请求
type UpdateContestRequest struct {
	Level               string                     `json:"level" binding:"omitempty,oneof=practice school cross_school platform"`
	Title               string                     `json:"title" binding:"omitempty,max=200"`
	Description         string                     `json:"description" binding:"omitempty"`
	Cover               string                     `json:"cover" binding:"omitempty,max=500"`
	Rules               string                     `json:"rules" binding:"omitempty"`
	StartTime           string                     `json:"start_time" binding:"omitempty"`
	EndTime             string                     `json:"end_time" binding:"omitempty"`
	RegistrationStart   string                     `json:"registration_start" binding:"omitempty"`
	RegistrationEnd     string                     `json:"registration_end" binding:"omitempty"`
	IsPublic            *bool                      `json:"is_public" binding:"omitempty"`
	MaxParticipants     *int                       `json:"max_participants" binding:"omitempty,min=0"`
	TeamMinSize         *int                       `json:"team_min_size" binding:"omitempty,min=1"`
	TeamMaxSize         *int                       `json:"team_max_size" binding:"omitempty,min=1"`
	DynamicScore        *bool                      `json:"dynamic_score" binding:"omitempty"`
	FirstBloodBonus     *int                       `json:"first_blood_bonus" binding:"omitempty,min=0"`
	Status              string                     `json:"status" binding:"omitempty,oneof=draft published ongoing ended"`
	BattleOrchestration *model.BattleOrchestration `json:"battle_orchestration" binding:"omitempty"`
}

// CreateChallengeRequest 创建题目请求
type CreateChallengeRequest struct {
	Title                  string                       `json:"title" binding:"required,max=200"`
	Description            string                       `json:"description" binding:"omitempty"`
	Category               string                       `json:"category" binding:"required,oneof=contract_vuln defi consensus crypto cross_chain nft reverse key_management misc"`
	RuntimeProfile         string                       `json:"runtime_profile" binding:"omitempty,oneof=static single_chain_instance fork_replay multi_service_lab"`
	Difficulty             int                          `json:"difficulty" binding:"omitempty,min=1,max=5"`
	BasePoints             int                          `json:"base_points" binding:"omitempty,min=1"`
	MinPoints              int                          `json:"min_points" binding:"omitempty,min=1"`
	DecayFactor            float64                      `json:"decay_factor" binding:"omitempty,min=0,max=1"`
	ContractCode           string                       `json:"contract_code" binding:"omitempty"`
	SetupCode              string                       `json:"setup_code" binding:"omitempty"`
	DeployScript           string                       `json:"deploy_script" binding:"omitempty"`
	CheckScript            string                       `json:"check_script" binding:"omitempty"`
	FlagType               string                       `json:"flag_type" binding:"omitempty,oneof=static dynamic"`
	FlagTemplate           string                       `json:"flag_template" binding:"omitempty,max=200"`
	Hints                  []ChallengeHintPayload       `json:"hints" binding:"omitempty,dive"`
	Attachments            []string                     `json:"attachments" binding:"omitempty,dive,max=500"`
	ChallengeOrchestration model.ChallengeOrchestration `json:"challenge_orchestration"`
	Tags                   []string                     `json:"tags" binding:"omitempty"`
	IsPublic               bool                         `json:"is_public"`
}

// UpdateChallengeRequest 更新题目请求
type UpdateChallengeRequest struct {
	Title                  *string                       `json:"title" binding:"omitempty,max=200"`
	Description            *string                       `json:"description" binding:"omitempty"`
	Category               *string                       `json:"category" binding:"omitempty,oneof=contract_vuln defi consensus crypto cross_chain nft reverse key_management misc"`
	RuntimeProfile         *string                       `json:"runtime_profile" binding:"omitempty,oneof=static single_chain_instance fork_replay multi_service_lab"`
	Difficulty             *int                          `json:"difficulty" binding:"omitempty,min=1,max=5"`
	BasePoints             *int                          `json:"base_points" binding:"omitempty,min=1"`
	MinPoints              *int                          `json:"min_points" binding:"omitempty,min=1"`
	DecayFactor            *float64                      `json:"decay_factor" binding:"omitempty,min=0,max=1"`
	ContractCode           *string                       `json:"contract_code" binding:"omitempty"`
	SetupCode              *string                       `json:"setup_code" binding:"omitempty"`
	DeployScript           *string                       `json:"deploy_script" binding:"omitempty"`
	CheckScript            *string                       `json:"check_script" binding:"omitempty"`
	FlagType               *string                       `json:"flag_type" binding:"omitempty,oneof=static dynamic"`
	FlagTemplate           *string                       `json:"flag_template" binding:"omitempty,max=200"`
	Hints                  *[]ChallengeHintPayload       `json:"hints" binding:"omitempty,dive"`
	Attachments            *[]string                     `json:"attachments" binding:"omitempty,dive,max=500"`
	ChallengeOrchestration *model.ChallengeOrchestration `json:"challenge_orchestration" binding:"omitempty"`
	Tags                   *[]string                     `json:"tags" binding:"omitempty"`
	IsPublic               *bool                         `json:"is_public" binding:"omitempty"`
	Status                 *string                       `json:"status" binding:"omitempty,oneof=draft active archived"`
}

// AddChallengeToContestRequest 添加题目到竞赛请求
type AddChallengeToContestRequest struct {
	ChallengeID uint `json:"challenge_id" binding:"required"`
	Points      int  `json:"points" binding:"omitempty,min=1"`
	SortOrder   int  `json:"sort_order" binding:"omitempty,min=0"`
	IsVisible   bool `json:"is_visible"`
}

// UpdateContestChallengeRequest 更新竞赛题目请求
type UpdateContestChallengeRequest struct {
	Points    *int  `json:"points" binding:"omitempty,min=1"`
	SortOrder *int  `json:"sort_order" binding:"omitempty,min=0"`
	IsVisible *bool `json:"is_visible" binding:"omitempty"`
}

// CreateTeamRequest 创建队伍请求
type CreateTeamRequest struct {
	ContestID   uint   `json:"contest_id" binding:"required"`
	Name        string `json:"name" binding:"required,max=100"`
	Avatar      string `json:"avatar" binding:"omitempty,max=500"`
	Description string `json:"description" binding:"omitempty"`
}

// UpdateTeamRequest 更新队伍请求
type UpdateTeamRequest struct {
	Name        string `json:"name" binding:"omitempty,max=100"`
	Avatar      string `json:"avatar" binding:"omitempty,max=500"`
	Description string `json:"description" binding:"omitempty"`
}

// JoinTeamRequest 加入队伍请求
type JoinTeamRequest struct {
	InviteCode string `json:"invite_code" binding:"required,max=50"`
}

type RequestChallengePublishRequest struct {
	Reason string `json:"reason" binding:"omitempty"`
}

// TransferLeaderRequest 转让队长请求
type TransferLeaderRequest struct {
	NewLeaderID uint `json:"new_leader_id" binding:"required"`
}

// KickMemberRequest 踢出成员请求
type KickMemberRequest struct {
	MemberID uint `json:"member_id" binding:"required"`
}

// SubmitFlagRequest 提交Flag请求
type SubmitFlagRequest struct {
	ChallengeID uint   `json:"challenge_id" binding:"required"`
	Flag        string `json:"flag" binding:"required,max=200"`
}

// StartChallengeEnvRequest 启动题目环境请求
type StartChallengeEnvRequest struct {
	ChallengeID uint `json:"challenge_id" binding:"required"`
}

// ListContestsRequest 竞赛列表请求
type ListContestsRequest struct {
	PaginationRequest
	Type     string `form:"type"`
	Level    string `form:"level"`
	Status   string `form:"status"`
	IsPublic *bool  `form:"is_public"`
	Keyword  string `form:"keyword"`
}

// ListChallengesRequest 题目列表请求
type ListChallengesRequest struct {
	PaginationRequest
	Category   string `form:"category"`
	Difficulty *int   `form:"difficulty" binding:"omitempty,min=1,max=5"`
	SourceType string `form:"source_type" binding:"omitempty,oneof=preset auto_converted user_created"`
	Status     string `form:"status"`
	IsPublic   *bool  `form:"is_public"`
	Keyword    string `form:"keyword"`
}

// ListTeamsRequest 队伍列表请求
type ListTeamsRequest struct {
	PaginationRequest
	ContestID uint   `form:"contest_id" binding:"required"`
	Status    string `form:"status"`
	Keyword   string `form:"keyword"`
}

// ListScoreboardRequest 排行榜请求
type ListScoreboardRequest struct {
	PaginationRequest
}

// InviteTeamMemberRequest 邀请队伍成员请求
type InviteTeamMemberRequest struct {
	UserID uint `json:"user_id" binding:"required"`
}
